### DISCUSSION TITLE
Migration Problems to Langium 2.0

### DISCUSSION THREAD

**Discussion by bjthehun** [eclipse-langium/langium#1172] [Q&A]
I'm having problems migrating to Langium 2.0 while using npm with Webpack as bundler. When I try to start the example project ![here](https://github.com/bjthehun/monaco-webpack) with `npm start`, I get five errors of the following kind:

```sh
ERROR in ./src/editql/generated/ast.ts 8:18-36
Module not found: Error: Package path . is not exported from package ~/monaco-webpack/node_modules/langium (see exports field in ~/monaco-webpack/node_modules/langium/package.json)
 @ ./src/editql/generated/module.ts 8:17-36
 @ ./src/editql/editql-module.ts 11:20-52
 @ ./src/editql/editql-lsp.ts 7:27-56
 @ ./src/index.tsx 165:14-168:6
```

Can you tell me what am I doing wrong?

**Comment by msujew** [comment]
Hey @bjthehun,

assuming you just updated to 2.0 without doing anything else, you probably ran into our ESM change. See [this changelog](https://github.com/eclipse-langium/langium/blob/main/packages/langium/CHANGELOG.md#ecmascript-modules-esm) entry. It contains tips on how to migrate to 2.0/ESM. If you're not able to change to ESM in your project, you can use Langium 1.3.0 for now. It contains the same feature set except for the ESM change.

**Reply by ballcoach12** [reply]
@bjthehun, I had the same issue earlier this week. I recommend running the latest version of the Yeoman generator and creating a new HelloWorld project and compare the 'package.json' file in your project to the one created by the generator.

After a few changes to my 'package.json' file, for me it was a matter of simply adding '.js' to the end of some import statements.

**Reply by bjthehun** [reply]
@msujew: I'm already using ESM and i've gone through the [migration manual linked in the changelog](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#how-can-i-make-my-typescript-project-output-esm). I've sinced updated Node in  `package.json` to `>=16`, though that didn't change anything.

@ballcoach12 I've also tried to create a new HelloWorld project with Yeoman 4.3.1. The only change I would say could be significant is `esbuild` instead of `webpack` for code bundling. (I don't think I could switch the project I'm working on to use `esbuild`, for that matter.)

**Reply by msujew** [reply]
I'm very much confused by the error. Clicking on the file linked in the error shows that we have a perfectly valid `"."` export defined in `langium`. I've just tried it with esbuild, which seems to pick this up just fine. TypeScript also doesn't have any issues with the import. I have absolutely no clue why the webpack build fails.

**Reply by bjthehun** [reply]
Update: I've been trying over the weekend to get Langium 2.0. running with `webpack`, and eventually figured out that the latter tool doesn't pick up the `import`  conditional exports for `.`. When I replaced this export with `default` (`chevrotain-allstar` also provides an `default` export), I got this follow-up error:

```sh
ERROR in ./node_modules/langium/lib/parser/langium-parser.js 7:0-61
Module not found: Error: Default condition should be last one
 @ ./node_modules/langium/lib/default-module.js 45:0-79 63:50-83
 @ ./node_modules/langium/lib/index.js 6:0-36 6:0-36
 @ ./src/editql/editql-lsp.ts 8:18-36
 @ ./src/index.tsx 165:14-168:6
```

After swapping the conditional exports in the `package.json` file for `chevrotain-allstar` like this:

```json
"exports": {
    ".": {
        "types": "./lib/index.d.ts",
        "default": "./lib/index.js"
    }
},
```

the `webpack` build succeeded for me.

So I would ask if the Langium exports could be changed to use `default` instead of  `import`. I'm not familiar enough with JavaScript bundlers so that I could rule out problems for other Langium users, however.

**Reply by davydkov** [reply]
Yes, and [PR #2](https://github.com/langium/chevrotain-allstar/pull/2) with the fix is waiting 2 weeks ago
